cmd/enterprise-portal/internal/database/importer/importer.go

```diff
@@ -0,0 +1,296 @@
+func (i *importer) importSubscription(ctx context.Context, dotcomSub *dotcomdb.SubscriptionAttributes) (err error) {
+	tr, ctx := trace.New(ctx, "importSubscription",
+		attribute.String("dotcomSub.ID", dotcomSub.ID))
+	defer tr.EndWithErr(&err)
+
+	dotcomLicenses, err := i.dotcom.ListEnterpriseSubscriptionLicenses(ctx, []*subscriptionsv1.ListEnterpriseSubscriptionLicensesFilter{{
+		Filter: &subscriptionsv1.ListEnterpriseSubscriptionLicensesFilter_SubscriptionId{
+			SubscriptionId: dotcomSub.ID,
+		},
+	}}, 0) // list all licenses
+	if err != nil {
+		return errors.Wrap(err, "dotcom: list licenses")
+	}
+
+	tr.SetAttributes(attribute.Int("licenses", len(dotcomLicenses)))
+	if len(dotcomLicenses) == 0 {
+		// We don't care about this subscription
+		return nil
+	}
+
+	// Construct the conditions we need to apply during the subscription
+	// update
+	var conditions []subscriptions.CreateSubscriptionConditionOptions
+	if epSub, err := i.subscriptions.Get(ctx, dotcomSub.ID); err == nil {
+		// No error - this subscription already exists.
+		tr.SetAttributes(attribute.Bool("already_imported", true))
+		if epSub.ArchivedAt == nil && dotcomSub.ArchivedAt != nil {
+			// This subscription was archived post-import. We need to
+			// add a new condition that reflects the archival event.
+			tr.SetAttributes(attribute.Bool("archived", true))
+			conditions = append(conditions,
+				subscriptions.CreateSubscriptionConditionOptions{
+					Status:         subscriptionsv1.EnterpriseSubscriptionCondition_STATUS_ARCHIVED,
+					TransitionTime: utctime.FromTime(*dotcomSub.ArchivedAt),
+				})
+		}
+	} else if !errors.Is(err, subscriptions.ErrSubscriptionNotFound) {
+		// Error occured, and it's not subscription-not-found - fail.
+		return errors.Wrap(err, "check for already-imported subscription")
+	} else {
+		// Subscription doesn't exist, so we need to create all the
+		// conditions that reflect the creation and archival events.
+		tr.SetAttributes(attribute.Bool("already_imported", false))
+		conditions = append(conditions,
+			subscriptions.CreateSubscriptionConditionOptions{
+				Status:         subscriptionsv1.EnterpriseSubscriptionCondition_STATUS_CREATED,
+				TransitionTime: utctime.FromTime(dotcomSub.CreatedAt),
+			})
+		if dotcomSub.ArchivedAt != nil {
+			tr.SetAttributes(attribute.Bool("archived", true))
+			conditions = append(conditions,
+				subscriptions.CreateSubscriptionConditionOptions{
+					Status:         subscriptionsv1.EnterpriseSubscriptionCondition_STATUS_ARCHIVED,
+					TransitionTime: utctime.FromTime(*dotcomSub.ArchivedAt),
+				})
+		}
+	}
+
+	// Apply updates to the subscription, creating it if it does not
+	// exist yet.
+	activeLicense := dotcomLicenses[0]
+	if _, err := i.subscriptions.Upsert(ctx, dotcomSub.ID,
+		subscriptions.UpsertSubscriptionOptions{
+			DisplayName: func() *sql.NullString {
+				for _, t := range activeLicense.Tags {
+					parts := strings.SplitN(t, ":", 2)
+					if len(parts) != 2 {
+						return nil
```

Comment:
Is the order of tags implementation defined to always return tags with `key:val` first?
Or should this keep iterating over all the tags rather than stopping if it encountered a tag like `trueup` before `customer:name`?
```suggestion
						continue
```

Looking at dotcom the `customer:<name>` tag always seems to come first so it's probably not an issue but I don't know what the expected behaviour is

Reply:
Ah, good catch! this should definitely be a `continue`